### PR TITLE
feat: rpc2: protocol generator

### PR DESCRIPTION
**Pull request by lucix-aws**
* Adds protocol codegen for `smithy.protocols#rpcv2Cbor`.
* Adds coercion utilities to `encoding/cbor` to aid deserialization.
* Add prioritized protocol selection (if you blink you'll miss it).
* I added and tested document support, but it's unexported for now until we enable them in the protocol.

Event streams are not supported here due to the dependency being locked up in the SDK repo. Protocol tests are also generated SDK-side, since that's where all of them are right now.

### CODE REVIEW

**Review by wty-Bryant — APPROVED** [review]

File: codegen/smithy-go-codegen/src/main/java/software/amazon/smithy/go/codegen/integration/DefaultProtocols.java [hunk 1]
```diff
@@ -0,0 +1,27 @@
...
+ * permissions and limitations under the License.
+ */
+
+package software.amazon.smithy.go.codegen.integration;
+
+import java.util.List;
+import software.amazon.smithy.go.codegen.protocol.rpc2.cbor.Rpc2CborProtocolGenerator;
+import software.amazon.smithy.utils.ListUtils;
+
+public class DefaultProtocols implements GoIntegration {
```
Comment (aajtodd) [inline codegen/smithy-go-codegen/src/main/java/software/amazon/smithy/go/codegen/integration/DefaultProtocols.java:22]: **question**: What makes this "default"?

File: codegen/smithy-go-codegen/src/main/java/software/amazon/smithy/go/codegen/protocol/rpc2/Rpc2ProtocolGenerator.java [hunk 2]
```diff
@@ -0,0 +1,64 @@
...
+import software.amazon.smithy.codegen.core.CodegenException;
+import software.amazon.smithy.go.codegen.ApplicationProtocol;
+import software.amazon.smithy.go.codegen.integration.ProtocolGenerator;
+import software.amazon.smithy.go.codegen.protocol.DeserializeResponseMiddleware;
+import software.amazon.smithy.model.knowledge.TopDownIndex;
+import software.amazon.smithy.model.shapes.OperationShape;
+import software.amazon.smithy.utils.SmithyInternalApi;
+
+@SmithyInternalApi
+public abstract class Rpc2ProtocolGenerator implements ProtocolGenerator {
```
Comment (aajtodd) [inline codegen/smithy-go-codegen/src/main/java/software/amazon/smithy/go/codegen/protocol/rpc2/Rpc2ProtocolGenerator.java:29]: **naming**: Should this have `cbor` somewhere in the name?

**Review by aajtodd — APPROVED** [review]
Structurally this looks fine, no concerns.

File: codegen/smithy-go-codegen/src/main/java/software/amazon/smithy/go/codegen/protocol/rpc2/Rpc2ProtocolGenerator.java (hunk 2 above)
Comment (lucix-aws) [reply codegen/smithy-go-codegen/src/main/java/software/amazon/smithy/go/codegen/protocol/rpc2/Rpc2ProtocolGenerator.java:29]: `Rpc2CborProtocolGenerator` inherits from this.

**Pull request merged**